source/opt/propagator.cpp

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "propagator.h"
+
+namespace spvtools {
+namespace opt {
+
+template <typename ValueType>
+void SSAPropagator<ValueType>::AddControlEdge(const Edge& edge) {
+  const ir::BasicBlock *dest_bb = edge.dst;
+
+  // Refuse to add the exit block to the work list.
+  if (dest_bb == cfg_->pseudo_exit_block()) {
+    return;
+  }
+
+  // Try to mark the edge executable.  If it was already in the set of
+  // executable edges, do nothing.
+  if (MarkEdgeExecutable(edge)) {
+    return;
+  }
+
+  // If the edge had not already been marked executable, add the destination
+  // basic block to the work list.
+  blocks_.push(dest_bb);
+}
+
+template <typename ValueType>
+void SSAPropagator<ValueType>::AddSSAEdges(uint32_t id) {
+  for (const auto& use : get_def_use_mgr()->GetUses(id)) {
+    ir::Instruction* instr = use.inst;
+
+    // If the basic block for |instr| has not been simulated yet, do nothing.
+    if (!IsBlockSimulated(ctx_->get_instr_block(instr))) {
+      continue;
+    }
+
+    if (ShouldSimulateAgain(instr)) {
+      ssa_edges_.push(instr);
+    }
+  }
+}
+
+template <typename ValueType>
+void SSAPropagator<ValueType>::Simulate(ir::Instruction* instr) {
+  // Don't bother visiting instructions that should not be simulated again.
+  if (!ShouldSimulateAgain(instr)) {
+    return;
+  }
+
+  ir::BasicBlock* dest_bb = nullptr;
+  uint32_t result_id = 0;
+  PropStatus status = visit_fn_(instr, &dest_bb, &result_id);
```

Comment:
You pass in the address |result_id|, but |result_id| is not used afterward.  Why do you need it?

Reply:
Fixed.